### PR TITLE
backupccl: mark type descs as dropped if there is a failure in restore

### PR DESCRIPTION
**Pull request by arulajmani**
Previously, when cleaning up type descriptors from a failed restore, we
would directly delete the system.descriptor entry. This is bad because
writing directly to the system.descriptor table means we bypass the
descriptor collection, which doesn't know these descriptors have been
dropped.

The descriptor collection validates all uncommitted descriptors before
writing them. As the descriptor collection doesn't know type
descriptors have been dropped, validating cross references was always
bound to fail. Put another way, if a restore from a backup which
contained user defined types failed for any reason, we were bound to
require manual cleanup.

This patch fixes this problem by going through the descriptor
collection and writing the descriptor in DROPPED state in addition to
deleting the system.descriptor entry.

Release note (bug fix): Fixed a bug where a failed restore from a
backup including user defined types would require manual cleanup.

### CODE REVIEW

**Comment by arulajmani** [comment]
Found this when trying to investigate the cause of https://github.com/cockroachdb/cockroach/issues/60773 with @ajstorm

File: pkg/ccl/backupccl/backup_test.go [hunk 1]
```diff
@@ -6817,6 +6817,38 @@ func TestBackupDoesNotHangOnIntent(t *testing.T) {
...
+	sqlDB.Exec(t, `
+CREATE DATABASE db; 
+CREATE TYPE db.typ AS ENUM();
+CREATE TABLE db.table (k INT PRIMARY KEY, v db.typ);
+`)
+
+	// Back up the database, drop it, and restore into it.
+	sqlDB.Exec(t, `BACKUP DATABASE db TO 'nodelocal://0/test/'`)
+	sqlDB.Exec(t, `DROP DATABASE db`)
+	sqlDB.ExpectErr(t, "boom", `RESTORE DATABASE db FROM 'nodelocal://0/test/'`)
```
Comment (pbardea) [inline pkg/ccl/backupccl/backup_test.go:6849]: Can we add an assertion that the relevant descriptors are in the DROP state?

**Review by pbardea — APPROVED** [review]
Thank you! LGTM with a minor testing comment.

File: pkg/ccl/backupccl/backup_test.go (hunk 1 above)
Comment (arulajmani) [reply pkg/ccl/backupccl/backup_test.go:6849]: To clarify, they're only written in the DROP state for the purpose of the desc collection, we still have to overwrite this and delete the descriptor entry from system.descriptor (as before). 

Not quite sure how to test the descriptor entry has been removed successfully, as the ID is unknown and the namespace entry has been removed as well. Is there any existing tests I could take inspiration from?

**Review by ajwerner — APPROVED** [review]
I feel like this sort of thing should have been caught with testing at a different level. Like for all of the tests where we restore from real backups, it feels like we should inject failures at various stages and make sure that those all roll back correctly. For another PR. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62278#-:-MWRzDhv5MeXZE7pZc94:b-k6hc9l)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @pbardea)

___
*[pkg/ccl/backupccl/backup_test.go, line 6849 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62278#-MWBiDv1GU_TTvNq3U3L-r1-6849:-MWRyjRA5SYpvZejVh1K:b5d53tw) ([raw file](https://github.com/cockroachdb/cockroach/blob/9900c912fae45fd406ee54531b7f12c910c0869e/pkg/ccl/backupccl/backup_test.go#L6849)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

To clarify, they're only written in the DROP state for the purpose of the desc collection, we still have to overwrite this and delete the descriptor entry from system.descriptor (as before). 

Not quite sure how to test the descriptor entry has been removed successfully, as the ID is unknown and the namespace entry has been removed as well. Are there any existing tests I could take inspiration from?  
</blockquote></details>

Can you just check that `crdb_internal.invalid_objects` is empty? 

Maybe also validate that without your change that it isn't.

**Review by arulajmani — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62278)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @pbardea)

___
*[pkg/ccl/backupccl/backup_test.go, line 6849 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62278#-MWBiDv1GU_TTvNq3U3L-r1-6849:-MWUMnlT6BzUS0aJ81aN:b1ns6mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9900c912fae45fd406ee54531b7f12c910c0869e/pkg/ccl/backupccl/backup_test.go#L6849)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Can you just check that `crdb_internal.invalid_objects` is empty? 

Maybe also validate that without your change that it isn't. 
</blockquote></details>

`crdb_internal.invalid_objects` was empty without my changes as well because when the restore job failed it left behind all descriptor entries (not just the type descriptor), so the descriptor could be validated. I realized `system.namespace` is fair game though, so that's what I'm doing now.

**Comment by arulajmani** [comment]
TFTRs! 

bors r=ajwerner,pbardea

**Pull request merged**